# Capstone Project: Create a Customer Segmentation Report for Arvato Financial Services

In this project, you will analyze demographics data for customers of a mail-order sales company in Germany, comparing it against demographics information for the general population. You'll use unsupervised learning techniques to perform customer segmentation, identifying the parts of the population that best describe the core customer base of the company. Then, you'll apply what you've learned on a third dataset with demographics information for targets of a marketing campaign for the company, and use a model to predict which individuals are most likely to convert into becoming customers for the company. The data that you will use has been provided by our partners at Bertelsmann Arvato Analytics, and represents a real-life data science task.

If you completed the first term of this program, you will be familiar with the first part of this project, from the unsupervised learning project. The versions of those two datasets used in this project will include many more features and has not been pre-cleaned. You are also free to choose whatever approach you'd like to analyzing the data rather than follow pre-determined steps. In your work on this project, make sure that you carefully document your steps and decisions, since your main deliverable for this project will be a blog post reporting your findings.

In [1]:
# import libraries here; add more as necessary
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# magic word for producing visualizations in notebook
%matplotlib inline

## Part 0: Get to Know the Data

There are four data files associated with this project:

- `Udacity_AZDIAS_052018.csv`: Demographics data for the general population of Germany; 891 211 persons (rows) x 366 features (columns).
- `Udacity_CUSTOMERS_052018.csv`: Demographics data for customers of a mail-order company; 191 652 persons (rows) x 369 features (columns).
- `Udacity_MAILOUT_052018_TRAIN.csv`: Demographics data for individuals who were targets of a marketing campaign; 42 982 persons (rows) x 367 (columns).
- `Udacity_MAILOUT_052018_TEST.csv`: Demographics data for individuals who were targets of a marketing campaign; 42 833 persons (rows) x 366 (columns).

Each row of the demographics files represents a single person, but also includes information outside of individuals, including information about their household, building, and neighborhood. Use the information from the first two files to figure out how customers ("CUSTOMERS") are similar to or differ from the general population at large ("AZDIAS"), then use your analysis to make predictions on the other two files ("MAILOUT"), predicting which recipients are most likely to become a customer for the mail-order company.

The "CUSTOMERS" file contains three extra columns ('CUSTOMER_GROUP', 'ONLINE_PURCHASE', and 'PRODUCT_GROUP'), which provide broad information about the customers depicted in the file. The original "MAILOUT" file included one additional column, "RESPONSE", which indicated whether or not each recipient became a customer of the company. For the "TRAIN" subset, this column has been retained, but in the "TEST" subset it has been removed; it is against that withheld column that your final predictions will be assessed in the Kaggle competition.

Otherwise, all of the remaining columns are the same between the three data files. For more information about the columns depicted in the files, you can refer to two Excel spreadsheets provided in the workspace. [One of them](./DIAS Information Levels - Attributes 2017.xlsx) is a top-level list of attributes and descriptions, organized by informational category. [The other](./DIAS Attributes - Values 2017.xlsx) is a detailed mapping of data values for each feature in alphabetical order.

In the below cell, we've provided some initial code to load in the first two datasets. Note for all of the `.csv` data files in this project that they're semicolon (`;`) delimited, so an additional argument in the [`read_csv()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) call has been included to read in the data properly. Also, considering the size of the datasets, it may take some time for them to load completely.

You'll notice when the data is loaded in that a warning message will immediately pop up. Before you really start digging into the modeling and analysis, you're going to need to perform some cleaning. Take some time to browse the structure of the data and look over the informational spreadsheets to understand the data values. Make some decisions on which features to keep, which features to drop, and if any revisions need to be made on data formats. It'll be a good idea to create a function with pre-processing steps, since you'll need to clean all of the datasets before you work with them.

In [2]:
# load in the data
azdias = pd.read_csv('./data/Udacity_AZDIAS_052018.csv')
customers = pd.read_csv('./data/Udacity_CUSTOMERS_052018.csv')

# Ignore any empty columns while loading data
no_unamed = lambda x: 'Unnamed' not in x

attributes_values = pd.read_excel('./data/DIAS Attributes - Values 2017.xlsx', header=1, usecols=no_unamed)
attributes_info = pd.read_excel('./data/DIAS Information Levels - Attributes 2017.xlsx', header=1, usecols=no_unamed)

D:\Users\prana\Anaconda3\envs\Python3.7\lib\site-packages\IPython\core\interactiveshell.py:3051: DtypeWarning: Columns (18,19) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


The warning says columns 18 and 19 have mixed data types, we can explore these two columns in later steps. This is one point to remember.

### Checking data shapes:

Verifying whether shapes of dataframes and the extra columns that are there in the customers dataset as per the description.

In [3]:
print("Azdias shape -",azdias.shape)
print("Customers shape -",customers.shape)

Azdias shape - (891221, 366)
Customers shape - (191652, 369)


In [4]:
print("Number of columns in Customers Data - ", len(list(customers.columns)))
print("Number of columns in Azdius Data - ", len(list(azdias.columns)))

extra_cols_in_customers = [col for col in customers.columns if col not in azdias.columns]
print("Extra columns in Customers Data - ",extra_cols_in_customers)

Number of columns in Customers Data -  369
Number of columns in Azdius Data -  366
Extra columns in Customers Data -  ['PRODUCT_GROUP', 'CUSTOMER_GROUP', 'ONLINE_PURCHASE']


The shapes and extra columns match the description.

### Checkng heads of dataframes:

In [5]:
azdias.head()

LNR  AGER_TYP  AKT_DAT_KL  ALTER_HH  ALTER_KIND1  ALTER_KIND2  \
0  910215        -1         NaN       NaN          NaN          NaN   
1  910220        -1         9.0       0.0          NaN          NaN   
2  910225        -1         9.0      17.0          NaN          NaN   
3  910226         2         1.0      13.0          NaN          NaN   
4  910241        -1         1.0      20.0          NaN          NaN   

   ALTER_KIND3  ALTER_KIND4  ALTERSKATEGORIE_FEIN  ANZ_HAUSHALTE_AKTIV  ...  \
0          NaN          NaN                   NaN                  NaN  ...   
1          NaN          NaN                  21.0                 11.0  ...   
2          NaN          NaN                  17.0                 10.0  ...   
3          NaN          NaN                  13.0                  1.0  ...   
4          NaN          NaN                  14.0                  3.0  ...   

   VHN  VK_DHT4A  VK_DISTANZ  VK_ZG11  W_KEIT_KIND_HH  WOHNDAUER_2008  \
0  NaN       NaN         NaN      NaN             NaN             NaN   
1  4.0       8.0        11.0     10.0             3.0             9.0   
2  2.0       9.0         9.0      6.0             3.0             9.0   
3  0.0       7.0        10.0     11.0             NaN             9.0   
4  2.0       3.0         5.0      4.0             2.0             9.0   

   WOHNLAGE ZABEOTYP ANREDE_KZ ALTERSKATEGORIE_GROB  
0       NaN        3         1                    2  
1       4.0        5         2                    1  
2       2.0        5         2                    3  
3       7.0        3         2                    4  
4       3.0        4         1                    3  

[5 rows x 366 columns]

In [6]:
customers.head()

LNR  AGER_TYP  AKT_DAT_KL  ALTER_HH  ALTER_KIND1  ALTER_KIND2  \
0    9626         2         1.0      10.0          NaN          NaN   
1    9628        -1         9.0      11.0          NaN          NaN   
2  143872        -1         1.0       6.0          NaN          NaN   
3  143873         1         1.0       8.0          NaN          NaN   
4  143874        -1         1.0      20.0          NaN          NaN   

   ALTER_KIND3  ALTER_KIND4  ALTERSKATEGORIE_FEIN  ANZ_HAUSHALTE_AKTIV  ...  \
0          NaN          NaN                  10.0                  1.0  ...   
1          NaN          NaN                   NaN                  NaN  ...   
2          NaN          NaN                   0.0                  1.0  ...   
3          NaN          NaN                   8.0                  0.0  ...   
4          NaN          NaN                  14.0                  7.0  ...   

   VK_ZG11  W_KEIT_KIND_HH  WOHNDAUER_2008  WOHNLAGE  ZABEOTYP  \
0      2.0             6.0             9.0       7.0         3   
1      3.0             0.0             9.0       NaN         3   
2     11.0             6.0             9.0       2.0         3   
3      2.0             NaN             9.0       7.0         1   
4      4.0             2.0             9.0       3.0         1   

       PRODUCT_GROUP  CUSTOMER_GROUP ONLINE_PURCHASE ANREDE_KZ  \
0  COSMETIC_AND_FOOD     MULTI_BUYER               0         1   
1               FOOD    SINGLE_BUYER               0         1   
2  COSMETIC_AND_FOOD     MULTI_BUYER               0         2   
3           COSMETIC     MULTI_BUYER               0         1   
4               FOOD     MULTI_BUYER               0         1   

  ALTERSKATEGORIE_GROB  
0                    4  
1                    4  
2                    4  
3                    4  
4                    3  

[5 rows x 369 columns]

We can see there are a lot of missing values in both of the dataframes. Looking at the **extra columns** in the customers dataset.

In [7]:
customers[extra_cols_in_customers].head()

PRODUCT_GROUP CUSTOMER_GROUP  ONLINE_PURCHASE
0  COSMETIC_AND_FOOD    MULTI_BUYER                0
1               FOOD   SINGLE_BUYER                0
2  COSMETIC_AND_FOOD    MULTI_BUYER                0
3           COSMETIC    MULTI_BUYER                0
4               FOOD    MULTI_BUYER                0

These three columns have specific information about the customers and their buying behavior.  We also have two extra dataframes with attribute information i.e. the information about what each column represents.

### Attribute values information

Additionally we have been given two excel sheets which we have already loaded into `attribute_values` and `attribute_info`dataframes.

* `attribute_values` - has a description of columns in `azdias`and `customers` data, information about range of values each column can take and their respective meanings.
* `attribute_info` - has detailed information about each column with additional notes wherever required.

We can use these two dataframes to understand the data.

In [10]:
attributes_values.head(15)

Attribute                                   Description  Value  \
0               AGER_TYP                            best-ager typology     -1   
1                    NaN                                           NaN      0   
2                    NaN                                           NaN      1   
3                    NaN                                           NaN      2   
4                    NaN                                           NaN      3   
5   ALTERSKATEGORIE_GROB  age classification through prename analysis   -1, 0   
6                    NaN                                           NaN      1   
7                    NaN                                           NaN      2   
8                    NaN                                           NaN      3   
9                    NaN                                           NaN      4   
10                   NaN                                           NaN      9   
11              ALTER_HH                 main age within the household      0   
12                   NaN                                           NaN      1   
13                   NaN                                           NaN      2   
14                   NaN                                           NaN      3   

                             Meaning  
0                            unknown  
1         no classification possible  
2                    passive elderly  
3                   cultural elderly  
4          experience-driven elderly  
5                            unknown  
6                         < 30 years  
7                      30 - 45 years  
8                      46 - 60 years  
9                         > 60 years  
10             uniformly distributed  
11  unknown / no main age detectable  
12         01.01.1895 bis 31.12.1899  
13         01.01.1900 bis 31.12.1904  
14         01.01.1905 bis 31.12.1909

As seen above each attribute has multiple rows corresponding to different values that attribute can have. We can use pandas `ffill()` method to fill the nan values in **Attribute** column to easily access all the information regarding an attribute in one go.

In [11]:
attributes_values["Attribute"] = attributes_values["Attribute"].ffill()
attributes_values.head(10)

Attribute                                   Description  Value  \
0              AGER_TYP                            best-ager typology     -1   
1              AGER_TYP                                           NaN      0   
2              AGER_TYP                                           NaN      1   
3              AGER_TYP                                           NaN      2   
4              AGER_TYP                                           NaN      3   
5  ALTERSKATEGORIE_GROB  age classification through prename analysis   -1, 0   
6  ALTERSKATEGORIE_GROB                                           NaN      1   
7  ALTERSKATEGORIE_GROB                                           NaN      2   
8  ALTERSKATEGORIE_GROB                                           NaN      3   
9  ALTERSKATEGORIE_GROB                                           NaN      4   

                      Meaning  
0                     unknown  
1  no classification possible  
2             passive elderly  
3            cultural elderly  
4   experience-driven elderly  
5                     unknown  
6                  < 30 years  
7               30 - 45 years  
8               46 - 60 years  
9                  > 60 years

In [52]:
attributes_info.head(15)

Information level              Attribute  \
0                NaN               AGER_TYP   
1             Person   ALTERSKATEGORIE_GROB   
2                NaN              ANREDE_KZ   
3                NaN          CJT_GESAMTTYP   
4                NaN      FINANZ_MINIMALIST   
5                NaN          FINANZ_SPARER   
6                NaN       FINANZ_VORSORGER   
7                NaN         FINANZ_ANLEGER   
8                NaN  FINANZ_UNAUFFAELLIGER   
9                NaN       FINANZ_HAUSBAUER   
10               NaN              FINANZTYP   
11               NaN            GEBURTSJAHR   
12               NaN        GFK_URLAUBERTYP   
13               NaN       GREEN_AVANTGARDE   
14               NaN             HEALTH_TYP   

                                          Description  \
0                                  best-ager typology   
1                       age through prename analysis    
2                                              gender   
3   Customer-Journey-Typology relating to the pref...   
4          financial typology: low financial interest   
5                     financial typology: money saver   
6                     financial typology: be prepared   
7                        financial typology: investor   
8                    financial typology: unremarkable   
9     financial typology: main focus is the own house   
10       best descirbing financial type for the peron   
11                                      year of birth   
12                                    vacation habits   
13  the environmental sustainability is the domina...   
14                                    health typology   

                                     Additional notes  
0   in cooperation with Kantar TNS; the informatio...  
1   modelled on millions of first name-age-referen...  
2                                                 NaN  
3   relating to the preferred information, marketi...  
4   Gfk-Typology based on a representative househo...  
5                                                 NaN  
6                                                 NaN  
7                                                 NaN  
8                                                 NaN  
9                                                 NaN  
10                                                NaN  
11                                                NaN  
12                                                NaN  
13  own typology based on modelled on different AZ...  
14  in cooperation with Kantar TNS; the informatio...

### Addressing warning we got when loading data

Also, We can look at the datatypes of **columns 18 and 19** since we got a warning while we loaded the data.

In [14]:
print(azdias.columns[18:20])

warning_cols = list(azdias.columns[18:20])

Index(['CAMEO_DEUG_2015', 'CAMEO_INTL_2015'], dtype='object')


We can see what thes two columns represent from the attributes dataframe.

In [15]:
attributes_values[attributes_values["Attribute"] == "CAMEO_DEUG_2015"]

Attribute                             Description Value  \
51  CAMEO_DEUG_2015  CAMEO classification 2015 - Uppergroup    -1   
52  CAMEO_DEUG_2015                                     NaN     1   
53  CAMEO_DEUG_2015                                     NaN     2   
54  CAMEO_DEUG_2015                                     NaN     3   
55  CAMEO_DEUG_2015                                     NaN     4   
56  CAMEO_DEUG_2015                                     NaN     5   
57  CAMEO_DEUG_2015                                     NaN     6   
58  CAMEO_DEUG_2015                                     NaN     7   
59  CAMEO_DEUG_2015                                     NaN     8   
60  CAMEO_DEUG_2015                                     NaN     9   

                              Meaning  
51                            unknown  
52                        upper class  
53                  upper middleclass  
54           established middleclasse  
55  consumption-oriented middleclass   
56                active middleclass   
57        low-consumption middleclass  
58                  lower middleclass  
59                      working class  
60                urban working class

In [16]:
# "CAMEO_INTL_2015" ---> named as "CAMEO_DEUINTL_2015" in attribute values excel sheet
attributes_values[attributes_values["Attribute"] == "CAMEO_DEUINTL_2015"]

Attribute                                        Description  \
105  CAMEO_DEUINTL_2015  CAMEO classification 2015 - international typo...   
106  CAMEO_DEUINTL_2015  (each German CAMEO code belongs to one interna...   
107  CAMEO_DEUINTL_2015                                                NaN   
108  CAMEO_DEUINTL_2015                                                NaN   
109  CAMEO_DEUINTL_2015                                                NaN   
110  CAMEO_DEUINTL_2015                                                NaN   
111  CAMEO_DEUINTL_2015                                                NaN   
112  CAMEO_DEUINTL_2015                                                NaN   
113  CAMEO_DEUINTL_2015                                                NaN   
114  CAMEO_DEUINTL_2015                                                NaN   
115  CAMEO_DEUINTL_2015                                                NaN   
116  CAMEO_DEUINTL_2015                                                NaN   
117  CAMEO_DEUINTL_2015                                                NaN   
118  CAMEO_DEUINTL_2015                                                NaN   
119  CAMEO_DEUINTL_2015                                                NaN   
120  CAMEO_DEUINTL_2015                                                NaN   
121  CAMEO_DEUINTL_2015                                                NaN   
122  CAMEO_DEUINTL_2015                                                NaN   
123  CAMEO_DEUINTL_2015                                                NaN   
124  CAMEO_DEUINTL_2015                                                NaN   
125  CAMEO_DEUINTL_2015                                                NaN   
126  CAMEO_DEUINTL_2015                                                NaN   
127  CAMEO_DEUINTL_2015                                                NaN   
128  CAMEO_DEUINTL_2015                                                NaN   
129  CAMEO_DEUINTL_2015                                                NaN   
130  CAMEO_DEUINTL_2015                                                NaN   

    Value                                            Meaning  
105    -1                                            unknown  
106    11    Wealthy Households-Pre-Family Couples & Singles  
107    12     Wealthy Households-Young Couples With Children  
108    13  Wealthy Households-Families With School Age Ch...  
109    14  Wealthy Households-Older Families &  Mature Co...  
110    15            Wealthy Households-Elders In Retirement  
111    21  Prosperous Households-Pre-Family Couples & Sin...  
112    22  Prosperous Households-Young Couples With Children  
113    23  Prosperous Households-Families With School Age...  
114    24  Prosperous Households-Older Families & Mature ...  
115    25         Prosperous Households-Elders In Retirement  
116    31  Comfortable Households-Pre-Family Couples & Si...  
117    32  Comfortable Households-Young Couples With Chil...  
118    33  Comfortable Households-Families With School Ag...  
119    34  Comfortable Households-Older Families & Mature...  
120    35        Comfortable Households-Elders In Retirement  
121    41  Less Affluent Households-Pre-Family Couples & ...  
122    42  Less Affluent Households-Young Couples With Ch...  
123    43  Less Affluent Households-Families With School ...  
124    44  Less Affluent Households-Older Families & Matu...  
125    45      Less Affluent Households-Elders In Retirement  
126    51     Poorer Households-Pre-Family Couples & Singles  
127    52      Poorer Households-Young Couples With Children  
128    53  Poorer Households-Families With School Age Chi...  
129    54  Poorer Households-Older Families & Mature Couples  
130    55             Poorer Households-Elders In Retirement

Looking at above two dataframes, we can conclude that these two columns must have values in the range given in the "Value" column i.e. -1 for unknown and integer values for corresponding categories.

We can have a look at the values that are in the actual dataframes by printing the unique values in these columns.

In [17]:
azdias[warning_cols].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891221 entries, 0 to 891220
Data columns (total 2 columns):
 #   Column           Non-Null Count   Dtype 
---  ------           --------------   ----- 
 0   CAMEO_DEUG_2015  792242 non-null  object
 1   CAMEO_INTL_2015  792242 non-null  object
dtypes: object(2)
memory usage: 13.6+ MB


In [18]:
azdias[warning_cols[0]].unique()

array([nan, 8.0, 4.0, 2.0, 6.0, 1.0, 9.0, 5.0, 7.0, 3.0, '4', '3', '7',
       '2', '8', '9', '6', '5', '1', 'X'], dtype=object)

In [19]:
azdias[warning_cols[1]].unique()

array([nan, 51.0, 24.0, 12.0, 43.0, 54.0, 22.0, 14.0, 13.0, 15.0, 33.0,
       41.0, 34.0, 55.0, 25.0, 23.0, 31.0, 52.0, 35.0, 45.0, 44.0, 32.0,
       '22', '24', '41', '12', '54', '51', '44', '35', '23', '25', '14',
       '34', '52', '55', '31', '32', '15', '13', '43', '33', '45', 'XX'],
      dtype=object)

In [20]:
customers[warning_cols[0]].unique()

array([1.0, nan, 5.0, 4.0, 7.0, 3.0, 9.0, 2.0, 6.0, 8.0, '6', '3', '8',
       '9', '2', '4', '1', '7', '5', 'X'], dtype=object)

In [21]:
customers[warning_cols[1]].unique()

array([13.0, nan, 34.0, 24.0, 41.0, 23.0, 15.0, 55.0, 14.0, 22.0, 43.0,
       51.0, 33.0, 25.0, 44.0, 54.0, 32.0, 12.0, 35.0, 31.0, 45.0, 52.0,
       '45', '25', '55', '51', '14', '54', '43', '22', '15', '24', '35',
       '23', '12', '44', '41', '52', '31', '13', '34', '32', '33', 'XX'],
      dtype=object)

We have 'X' and 'XX' as values in these columns which have not been given in the description, also there are 'nan' values.

These values can be replaced with 'np.nan', as in the attributes dataframe it is described that any unknown values can be treated as '-1' but for simpicity we can consider this '-1' as 'np.nan'. Also there is inconsistency in the values i.e. some of them are strings and some of them are floats. We can to convert all of them into integers.

In [22]:
def fix_cameo_cols(df):
    """
    Function to replace 'X', 'XX' and 'nan' values with -1 in columns ['CAMEO_DEUG_2015', 'CAMEO_INTL_2015']
    """
    cols = ["CAMEO_DEUG_2015", "CAMEO_INTL_2015"]
    
    df[cols] = df[cols].replace({"X": np.nan, "XX": np.nan})
    df[cols] = df[cols].astype(float)
    
    return df

In [23]:
customers = fix_cameo_cols(customers)

azdias = fix_cameo_cols(azdias)

In [24]:
customers[warning_cols[1]].unique()

array([13., nan, 34., 24., 41., 23., 15., 55., 14., 22., 43., 51., 33.,
       25., 44., 54., 32., 12., 35., 31., 45., 52.])

In [25]:
azdias[warning_cols[1]].unique()

array([nan, 51., 24., 12., 43., 54., 22., 14., 13., 15., 33., 41., 34.,
       55., 25., 23., 31., 52., 35., 45., 44., 32.])

In [26]:
print(azdias.shape, customers.shape)

(891221, 366) (191652, 369)


As we can see now we dont have invalid entries in these two columns. Although we have 'nan' values which we are going to address in further analysis.


### What each column represents

Lets explore what each column represents and understand a bit more about the data. For this we are going to use the `attributes_values` dataframe since it has all the column names and their descriptions with corresponding values.

In [112]:
attributes_values.head(10)

Attribute                                   Description  Value  \
0              AGER_TYP                            best-ager typology     -1   
1              AGER_TYP                                           NaN      0   
2              AGER_TYP                                           NaN      1   
3              AGER_TYP                                           NaN      2   
4              AGER_TYP                                           NaN      3   
5  ALTERSKATEGORIE_GROB  age classification through prename analysis   -1, 0   
6  ALTERSKATEGORIE_GROB                                           NaN      1   
7  ALTERSKATEGORIE_GROB                                           NaN      2   
8  ALTERSKATEGORIE_GROB                                           NaN      3   
9  ALTERSKATEGORIE_GROB                                           NaN      4   

                      Meaning  
0                     unknown  
1  no classification possible  
2             passive elderly  
3            cultural elderly  
4   experience-driven elderly  
5                     unknown  
6                  < 30 years  
7               30 - 45 years  
8               46 - 60 years  
9                  > 60 years

Attribute info description 
* Attribute - is the column name.
* Description - containes a short description about what that column represents.
* Value - range of values that are recorded during data collection
* Meaning - meaning of corresponding value

We can have a look at all the attributes, and their description. Also at the same time what are the values recorded for unknown observations of each attribute. This will help us clean our data i.e. replacing all the values which are unknown with np.nan like we did with the columns 18, 19.  

In [125]:
pd.set_option("display.max_rows", 500)
attributes_values.dropna()

Attribute  \
0                        AGER_TYP   
5            ALTERSKATEGORIE_GROB   
11                       ALTER_HH   
33                      ANREDE_KZ   
36            ANZ_HAUSHALTE_AKTIV   
37                   ANZ_HH_TITEL   
38                   ANZ_PERSONEN   
39                      ANZ_TITEL   
40                       BALLRAUM   
48                       BIP_FLAG   
51                CAMEO_DEUG_2015   
61                 CAMEO_DEU_2015   
105            CAMEO_DEUINTL_2015   
106            CAMEO_DEUINTL_2015   
131                 CJT_GESAMTTYP   
138             D19_BANKEN_ANZ_12   
145             D19_BANKEN_ANZ_24   
152              D19_BANKEN_DATUM   
162          D19_BANKEN_DIREKT_RZ   
163          D19_BANKEN_DIREKT_RZ   
170           D19_BANKEN_GROSS_RZ   
171           D19_BANKEN_GROSS_RZ   
178           D19_BANKEN_LOKAL_RZ   
179           D19_BANKEN_LOKAL_RZ   
186      D19_BANKEN_OFFLINE_DATUM   
196       D19_BANKEN_ONLINE_DATUM   
206    D19_BANKEN_ONLINE_QUOTE_12   
217            D19_BANKEN_REST_RZ   
218            D19_BANKEN_REST_RZ   
225         D19_BEKLEIDUNG_GEH_RZ   
226         D19_BEKLEIDUNG_GEH_RZ   
233        D19_BEKLEIDUNG_REST_RZ   
234        D19_BEKLEIDUNG_REST_RZ   
241                D19_BILDUNG_RZ   
242                D19_BILDUNG_RZ   
249               D19_BIO_OEKO_RZ   
250               D19_BIO_OEKO_RZ   
257                   D19_BUCH_RZ   
258                   D19_BUCH_RZ   
265             D19_DIGIT_SERV_RZ   
266             D19_DIGIT_SERV_RZ   
273        D19_DROGERIEARTIKEL_RZ   
274        D19_DROGERIEARTIKEL_RZ   
281                D19_ENERGIE_RZ   
282                D19_ENERGIE_RZ   
289               D19_FREIZEIT_RZ   
290               D19_FREIZEIT_RZ   
297                 D19_GARTEN_RZ   
298                 D19_GARTEN_RZ   
305             D19_GESAMT_ANZ_12   
312             D19_GESAMT_ANZ_24   
319              D19_GESAMT_DATUM   
329      D19_GESAMT_OFFLINE_DATUM   
339       D19_GESAMT_ONLINE_DATUM   
349    D19_GESAMT_ONLINE_QUOTE_12   
360               D19_HANDWERK_RZ   
361               D19_HANDWERK_RZ   
368              D19_HAUS_DEKO_RZ   
369              D19_HAUS_DEKO_RZ   
376          D19_KINDERARTIKEL_RZ   
377          D19_KINDERARTIKEL_RZ   
384                 D19_KONSUMTYP   
391              D19_KK_KUNDENTYP   
398               D19_KOSMETIK_RZ   
399               D19_KOSMETIK_RZ   
406           D19_LEBENSMITTEL_RZ   
407           D19_LEBENSMITTEL_RZ   
414                  D19_LOTTO_RZ   
415                  D19_LOTTO_RZ   
422     D19_NAHRUNGSERGAENZUNG_RZ   
423     D19_NAHRUNGSERGAENZUNG_RZ   
430               D19_RATGEBER_RZ   
431               D19_RATGEBER_RZ   
438                 D19_REISEN_RZ   
439                 D19_REISEN_RZ   
446          D19_SAMMELARTIKEL_RZ   
447          D19_SAMMELARTIKEL_RZ   
454                 D19_SCHUHE_RZ   
455                 D19_SCHUHE_RZ   
462               D19_SONSTIGE_RZ   
463               D19_SONSTIGE_RZ   
470                D19_TECHNIK_RZ   
471                D19_TECHNIK_RZ   
478              D19_TELKO_ANZ_12   
485              D19_TELKO_ANZ_24   
492               D19_TELKO_DATUM   
502           D19_TELKO_MOBILE_RZ   
503           D19_TELKO_MOBILE_RZ   
510       D19_TELKO_OFFLINE_DATUM   
520        D19_TELKO_ONLINE_DATUM   
530             D19_TELKO_REST_RZ   
531             D19_TELKO_REST_RZ   
538            D19_TIERARTIKEL_RZ   
539            D19_TIERARTIKEL_RZ   
546            D19_VERSAND_ANZ_12   
553            D19_VERSAND_ANZ_24   
560             D19_VERSAND_DATUM   
570     D19_VERSAND_OFFLINE_DATUM   
580      D19_VERSAND_ONLINE_DATUM   
590   D19_VERSAND_ONLINE_QUOTE_12   
601           D19_VERSAND_REST_RZ   
602           D19_VERSAND_REST_RZ   
609         D19_VERSICHERUNGEN_RZ   
610         D19_VERSICHERUNGEN_RZ   
617              D19_VERSI_ANZ_12   
624              D19_VERSI_ANZ_24   
631          D19_VOLLSORTIMENT_RZ   
632          D19_VOLLSORTIME

By observing above dataframes, we can conclude that

* Not all the features have unknown values. Some of them have values specified and might not have nan values in the collected data. This means the features with no unkown values might be compulsory fields in a form.

* We can replace all the **unkown** values with 'np.nan' values. 

We can have a look at how many columns all the data frames have in common and how many columns are specific to each dataframe.

In [31]:
attributes_list  = attributes_values.Attribute.unique().tolist()
azdias_list = list(azdias.columns)
customers_list = list(customers.columns)

In [33]:
common_to_all = (set(attributes_list) & set(azdias_list) & set(customers_list))
unique_to_azdias = (set(azdias_list) - set(attributes_list) - set(customers_list))
unique_to_customers = (set(customers_list) - set(attributes_list) - set(azdias_list))
unique_to_attributes = (set(attributes_list) - set(customers_list) - set(azdias_list))
unique_to_attributes_vs_azdias = (set(attributes_list) - set(azdias_list))
unique_to_azdias_vs_attributes = (set(attributes_list) - set(azdias_list))
common_azdias_attributes = (set(azdias_list) & set(attributes_list))

print("No of columns common to all 3 daframes: ", str(len(common_to_all)))
print("No of columns specific to azdias: " + str(len(unique_to_azdias)))
print("No of columns specific to customers: " + str(len(unique_to_customers)))
print("No of columns specific to attributes: " + str(len(unique_to_attributes)))
print("No of columns common between azdias and attributes: " + str(len(common_azdias_attributes)))
print("No of columns exclusive to attributes vs azdias: " + str(len(unique_to_attributes_vs_azdias)))
print("No of columns exclusive to azdias vs attributes: " + str(len(unique_to_azdias_vs_attributes)))

No of columns common to all 3 daframes:  272
No of columns specific to azdias: 0
No of columns specific to customers: 3
No of columns specific to attributes: 42
No of columns common between azdias and attributes: 272
No of columns exclusive to attributes vs azdias: 42
No of columns exclusive to azdias vs attributes: 42


* We have 42 unique columns given in the attribute-values data which are not present in azdias or customer data. This might be due to errors in recording these columns. We can go further correcting these mistakes and analysing what is the reason behing these missing columns. But for this analysis we are going to neglect this information and consider only the 272 columns which are common to all three dataframes to replace the 'unkown' values.

In [77]:
# unkown_attributes_values = attributes_values[(attributes_values["Meaning"] == "unknown") | ("no" in attributes_values["Meaning"])]

unkown_attributes_values = attributes_values[attributes_values["Meaning"] == "unknown"]
print(len(unkown_attributes_values))

232


We have **232** columns which have unkoen values.

In [78]:
unkown_attributes_values

Attribute  \
0                        AGER_TYP   
5            ALTERSKATEGORIE_GROB   
33                      ANREDE_KZ   
40                       BALLRAUM   
48                       BIP_FLAG   
51                CAMEO_DEUG_2015   
105            CAMEO_DEUINTL_2015   
131                 CJT_GESAMTTYP   
391              D19_KK_KUNDENTYP   
647                      EWDICHTE   
654                     FINANZTYP   
661                FINANZ_ANLEGER   
667              FINANZ_HAUSBAUER   
673             FINANZ_MINIMALIST   
679                 FINANZ_SPARER   
685         FINANZ_UNAUFFAELLIGER   
691              FINANZ_VORSORGER   
697                   GEBAEUDETYP   
724                 GEOSCORE_KLS7   
734             HAUSHALTSSTRUKTUR   
744                    HEALTH_TYP   
749            HH_EINKOMMEN_SCORE   
756                    INNENSTADT   
765                  KBA05_ALTER1   
771                  KBA05_ALTER2   
777                  KBA05_ALTER3   
783                  KBA05_ALTER4   
790                  KBA05_ANHANG   
795                   KBA05_ANTG1   
801                   KBA05_ANTG2   
807                   KBA05_ANTG3   
812                   KBA05_ANTG4   
821                KBA05_AUTOQUOT   
822                  KBA05_BAUMAX   
828                    KBA05_CCM1   
834                    KBA05_CCM2   
840                    KBA05_CCM3   
846                    KBA05_CCM4   
852                  KBA05_DIESEL   
858                    KBA05_FRAU   
864                     KBA05_GBZ   
870                  KBA05_HERST1   
877                  KBA05_HERST2   
884                  KBA05_HERST3   
891                  KBA05_HERST4   
898                  KBA05_HERST5   
905               KBA05_HERSTTEMP   
911                KBA05_KRSAQUOT   
917               KBA05_KRSHERST1   
923               KBA05_KRSHERST2   
929               KBA05_KRSHERST3   
935                KBA05_KRSKLEIN   
939                 KBA05_KRSOBER   
943                  KBA05_KRSVAN   
947                  KBA05_KRSZUL   
951                     KBA05_KW1   
957                     KBA05_KW2   
963                     KBA05_KW3   
969                   KBA05_MAXAH   
975                   KBA05_MAXBJ   
980                KBA05_MAXHERST   
986                  KBA05_MAXSEG   
991                 KBA05_MAXVORB   
995                    KBA05_MOD1   
1001                   KBA05_MOD2   
1007                   KBA05_MOD3   
1013                   KBA05_MOD4   
1020                   KBA05_MOD8   
1025                KBA05_MODTEMP   
1031                  KBA05_MOTOR   
1036                 KBA05_MOTRAD   
1041                   KBA05_SEG1   
1046                  KBA05_SEG10   
1052                   KBA05_SEG2   
1058                   KBA05_SEG3   
1064                   KBA05_SEG4   
1070                   KBA05_SEG5   
1076                   KBA05_SEG6   
1079                   KBA05_SEG7   
1084                   KBA05_SEG8   
1089                   KBA05_SEG9   
1094                  KBA05_VORB0   
1100                  KBA05_VORB1   
1106                  KBA05_VORB2   
1113                   KBA05_ZUL1   
1119                   KBA05_ZUL2   
1125                   KBA05_ZUL3   
1132                   KBA05_ZUL4   
1139         KBA13_ALTERHALTER_30   
1146         KBA13_ALTERHALTER_45   
1153         KBA13_ALTERHALTER_60   
1160         KBA13_ALTERHALTER_61   
1168                   KBA13_AUDI   
1175              KBA13_AUTOQUOTE   
1182                KBA13_BJ_1999   
1189                KBA13_BJ_2000   
1196                KBA13_BJ_2004   
1203                KBA13_BJ_2006   
1210                KBA13_BJ_2008   
1217                KBA13_BJ_2009   
1224                    KBA13_BMW   
1231               KBA13_CCM_1000   
1238               KBA13_CCM_1200   
1245               KBA13_CCM_1400   
1252             KBA13_CCM_0_1400   
1259               KBA13_CCM_1500   
1266          KBA13_CCM_1400_2500   
1273               KBA13_CCM

In [79]:
def get_unkown_repr(attrib):
    """Returns the representation of unkown values"""
    
    unkown = unkown_attributes_values[unkown_attributes_values["Attribute"] == attrib]["Value"]
    unkown = unkown.astype(str).str.cat(sep=",")
    unkown = [int(x) for x in unkown.split(",")]
      
    return [unkown]

def replace_unkown_with_nan(val, unkown):
    """Replaces unkown values with np.nan values if the value mathes with the unkown representation"""
    if val in unkown:
        return np.nan
    else:
        return val

In [82]:
ager_typ_unkwon = get_unkown_repr("AGER_TYP")

print(ager_typ_unkwon)

[[-1]]


lets have a look at azdias dataframe before replacing all the 'unkown' values. 

In [68]:
azdias.head(15)

LNR  AGER_TYP  AKT_DAT_KL  ALTER_HH  ALTER_KIND1  ALTER_KIND2  \
0   910215        -1         NaN       NaN          NaN          NaN   
1   910220        -1         9.0       0.0          NaN          NaN   
2   910225        -1         9.0      17.0          NaN          NaN   
3   910226         2         1.0      13.0          NaN          NaN   
4   910241        -1         1.0      20.0          NaN          NaN   
5   910244         3         1.0      10.0          NaN          NaN   
6   910248        -1         9.0       0.0          NaN          NaN   
7   910261        -1         1.0      14.0          NaN          NaN   
8   645145        -1         9.0      16.0          NaN          NaN   
9   645153        -1         5.0      17.0          NaN          NaN   
10  645165         0         1.0      10.0          NaN          NaN   
11  645169        -1         NaN       NaN          NaN          NaN   
12  612558        -1         5.0      21.0          NaN          NaN   
13  612561        -1         8.0      20.0          NaN          NaN   
14  612565        -1         NaN       NaN          NaN          NaN   

    ALTER_KIND3  ALTER_KIND4  ALTERSKATEGORIE_FEIN  ANZ_HAUSHALTE_AKTIV  ...  \
0           NaN          NaN                   NaN                  NaN  ...   
1           NaN          NaN                  21.0                 11.0  ...   
2           NaN          NaN                  17.0                 10.0  ...   
3           NaN          NaN                  13.0                  1.0  ...   
4           NaN          NaN                  14.0                  3.0  ...   
5           NaN          NaN                  10.0                  5.0  ...   
6           NaN          NaN                   NaN                  4.0  ...   
7           NaN          NaN                  14.0                  6.0  ...   
8           NaN          NaN                  16.0                  2.0  ...   
9           NaN          NaN                  17.0                  9.0  ...   
10          NaN          NaN                  10.0                  6.0  ...   
11          NaN          NaN                   NaN                  NaN  ...   
12          NaN          NaN                  14.0                  1.0  ...   
13          NaN          NaN                  20.0                  2.0  ...   
14          NaN          NaN                   NaN                  NaN  ...   

    VHN  VK_DHT4A  VK_DISTANZ  VK_ZG11  W_KEIT_KIND_HH  WOHNDAUER_2008  \
0   NaN       NaN         NaN      NaN             NaN             NaN   
1   4.0       8.0        11.0     10.0             3.0             9.0   
2   2.0       9.0         9.0      6.0             3.0             9.0   
3   0.0       7.0        10.0     11.0             NaN             9.0   
4   2.0       3.0         5.0      4.0             2.0             9.0   
5   2.0      10.0         7.0      4.0             6.0             9.0   
6   2.0       7.0        10.0     10.0             3.0             9.0   
7   2.0      10.0        12.0      9.0             5.0             9.0   
8   4.0       8.0        11.0      8.0             5.0             8.0   
9   4.0       1.0         1.0      1.0             4.0             3.0   
10  4.0       8.0        11.0     10.0             6.0             9.0   
11  NaN       NaN         NaN      NaN             NaN             NaN   
12  4.0       3.0         5.0      3.0             6.0             4.0   
13  NaN       3.0         6.0      8.0             6.0             3.0   
14  NaN       NaN         NaN      NaN             NaN             NaN   

    WOHNLAGE ZABEOTYP  ANREDE_KZ  ALTERSKATEGORIE_GROB  
0        NaN        3          1                     2  
1        4.0        5          2                     1  
2        2.0        5          2                     3  
3        7.0        3          2                     4  
4        3.0        4          1                     3  
5        7.0        4          2              

Looking at second column `AGER_TYP` there are a lot of *-1*s which correspond to **unkown** values. To convert all these values into `np.nan` values we first have to get the **value** which corresponds to unkown value from the `attribute_value` data and then replace these values. 

In [85]:
azdias["AGER_TYP"] = azdias["AGER_TYP"].apply(replace_unkown_with_nan, args=(ager_typ_unkwon))

In [86]:
azdias.head(15)

LNR  AGER_TYP  AKT_DAT_KL  ALTER_HH  ALTER_KIND1  ALTER_KIND2  \
0   910215       NaN         NaN       NaN          NaN          NaN   
1   910220       NaN         9.0       0.0          NaN          NaN   
2   910225       NaN         9.0      17.0          NaN          NaN   
3   910226       2.0         1.0      13.0          NaN          NaN   
4   910241       NaN         1.0      20.0          NaN          NaN   
5   910244       3.0         1.0      10.0          NaN          NaN   
6   910248       NaN         9.0       0.0          NaN          NaN   
7   910261       NaN         1.0      14.0          NaN          NaN   
8   645145       NaN         9.0      16.0          NaN          NaN   
9   645153       NaN         5.0      17.0          NaN          NaN   
10  645165       0.0         1.0      10.0          NaN          NaN   
11  645169       NaN         NaN       NaN          NaN          NaN   
12  612558       NaN         5.0      21.0          NaN          NaN   
13  612561       NaN         8.0      20.0          NaN          NaN   
14  612565       NaN         NaN       NaN          NaN          NaN   

    ALTER_KIND3  ALTER_KIND4  ALTERSKATEGORIE_FEIN  ANZ_HAUSHALTE_AKTIV  ...  \
0           NaN          NaN                   NaN                  NaN  ...   
1           NaN          NaN                  21.0                 11.0  ...   
2           NaN          NaN                  17.0                 10.0  ...   
3           NaN          NaN                  13.0                  1.0  ...   
4           NaN          NaN                  14.0                  3.0  ...   
5           NaN          NaN                  10.0                  5.0  ...   
6           NaN          NaN                   NaN                  4.0  ...   
7           NaN          NaN                  14.0                  6.0  ...   
8           NaN          NaN                  16.0                  2.0  ...   
9           NaN          NaN                  17.0                  9.0  ...   
10          NaN          NaN                  10.0                  6.0  ...   
11          NaN          NaN                   NaN                  NaN  ...   
12          NaN          NaN                  14.0                  1.0  ...   
13          NaN          NaN                  20.0                  2.0  ...   
14          NaN          NaN                   NaN                  NaN  ...   

    VHN  VK_DHT4A  VK_DISTANZ  VK_ZG11  W_KEIT_KIND_HH  WOHNDAUER_2008  \
0   NaN       NaN         NaN      NaN             NaN             NaN   
1   4.0       8.0        11.0     10.0             3.0             9.0   
2   2.0       9.0         9.0      6.0             3.0             9.0   
3   0.0       7.0        10.0     11.0             NaN             9.0   
4   2.0       3.0         5.0      4.0             2.0             9.0   
5   2.0      10.0         7.0      4.0             6.0             9.0   
6   2.0       7.0        10.0     10.0             3.0             9.0   
7   2.0      10.0        12.0      9.0             5.0             9.0   
8   4.0       8.0        11.0      8.0             5.0             8.0   
9   4.0       1.0         1.0      1.0             4.0             3.0   
10  4.0       8.0        11.0     10.0             6.0             9.0   
11  NaN       NaN         NaN      NaN             NaN             NaN   
12  4.0       3.0         5.0      3.0             6.0             4.0   
13  NaN       3.0         6.0      8.0             6.0             3.0   
14  NaN       NaN         NaN      NaN             NaN             NaN   

    WOHNLAGE ZABEOTYP  ANREDE_KZ  ALTERSKATEGORIE_GROB  
0        NaN        3          1                     2  
1        4.0        5          2                     1  
2        2.0        5          2                     3  
3        7.0        3          2                     4  
4        3.0        4          1                     3  
5        7.0        4          2              

We can repeat the same thing for all the 232 columns with unkwon values.

In [89]:
def replace_unkowns(df, unkown_attributes_values, verbose=False):
    """
    Replaces unkown values to 'np.nan' in all the columns provided in unkown_attributes_values list.
    """
    for attrib in unkown_attributes_values.Attribute:
        unkown = get_unkown_repr(attrib)
        if verbose:
            print("Replacing {} to NaN in Attribute {}".format(unkown, attrib))
        if attrib in df.columns:
            df[attrib] = df[attrib].apply(replace_unkown_with_nan, args=(unkown))
    return df

In [88]:
azdias = replace_unkowns(azdias, unkown_attributes_values, verbose=True)

Replacing [[-1]] to NaN in Attribute AGER_TYP
Replacing [[-1, 0]] to NaN in Attribute ALTERSKATEGORIE_GROB
Replacing [[-1, 0]] to NaN in Attribute ANREDE_KZ
Replacing [[-1]] to NaN in Attribute BALLRAUM
Replacing [[-1]] to NaN in Attribute BIP_FLAG
Replacing [[-1]] to NaN in Attribute CAMEO_DEUG_2015
Replacing [[-1]] to NaN in Attribute CAMEO_DEUINTL_2015
Replacing [[0]] to NaN in Attribute CJT_GESAMTTYP
Replacing [[-1]] to NaN in Attribute D19_KK_KUNDENTYP
Replacing [[-1]] to NaN in Attribute EWDICHTE
Replacing [[-1]] to NaN in Attribute FINANZTYP
Replacing [[-1]] to NaN in Attribute FINANZ_ANLEGER
Replacing [[-1]] to NaN in Attribute FINANZ_HAUSBAUER
Replacing [[-1]] to NaN in Attribute FINANZ_MINIMALIST
Replacing [[-1]] to NaN in Attribute FINANZ_SPARER
Replacing [[-1]] to NaN in Attribute FINANZ_UNAUFFAELLIGER
Replacing [[-1]] to NaN in Attribute FINANZ_VORSORGER
Replacing [[-1, 0]] to NaN in Attribute GEBAEUDETYP
Replacing [[-1, 0]] to NaN in Attribute GEOSCORE_KLS7
Replacing [[-1

Replacing [[-1]] to NaN in Attribute KBA13_KW_70
Replacing [[-1]] to NaN in Attribute KBA13_KW_61_120
Replacing [[-1]] to NaN in Attribute KBA13_KW_80
Replacing [[-1]] to NaN in Attribute KBA13_KW_90
Replacing [[-1]] to NaN in Attribute KBA13_KW_110
Replacing [[-1]] to NaN in Attribute KBA13_KW_120
Replacing [[-1]] to NaN in Attribute KBA13_KW_121
Replacing [[-1]] to NaN in Attribute KBA13_MAZDA
Replacing [[-1]] to NaN in Attribute KBA13_MERCEDES
Replacing [[-1]] to NaN in Attribute KBA13_MOTOR
Replacing [[-1]] to NaN in Attribute KBA13_NISSAN
Replacing [[-1]] to NaN in Attribute KBA13_OPEL
Replacing [[-1]] to NaN in Attribute KBA13_PEUGEOT
Replacing [[-1]] to NaN in Attribute KBA13_RENAULT
Replacing [[-1]] to NaN in Attribute KBA13_SEG_GELAENDEWAGEN
Replacing [[-1]] to NaN in Attribute KBA13_SEG_GROSSRAUMVANS
Replacing [[-1]] to NaN in Attribute KBA13_SEG_KLEINST
Replacing [[-1]] to NaN in Attribute KBA13_SEG_KLEINWAGEN
Replacing [[-1]] to NaN in Attribute KBA13_SEG_KOMPAKTKLASSE
Repl

In [90]:
customers = replace_unkowns(customers, unkown_attributes_values, verbose=True)

Replacing [[-1]] to NaN in Attribute AGER_TYP
Replacing [[-1, 0]] to NaN in Attribute ALTERSKATEGORIE_GROB
Replacing [[-1, 0]] to NaN in Attribute ANREDE_KZ
Replacing [[-1]] to NaN in Attribute BALLRAUM
Replacing [[-1]] to NaN in Attribute BIP_FLAG
Replacing [[-1]] to NaN in Attribute CAMEO_DEUG_2015
Replacing [[-1]] to NaN in Attribute CAMEO_DEUINTL_2015
Replacing [[0]] to NaN in Attribute CJT_GESAMTTYP
Replacing [[-1]] to NaN in Attribute D19_KK_KUNDENTYP
Replacing [[-1]] to NaN in Attribute EWDICHTE
Replacing [[-1]] to NaN in Attribute FINANZTYP
Replacing [[-1]] to NaN in Attribute FINANZ_ANLEGER
Replacing [[-1]] to NaN in Attribute FINANZ_HAUSBAUER
Replacing [[-1]] to NaN in Attribute FINANZ_MINIMALIST
Replacing [[-1]] to NaN in Attribute FINANZ_SPARER
Replacing [[-1]] to NaN in Attribute FINANZ_UNAUFFAELLIGER
Replacing [[-1]] to NaN in Attribute FINANZ_VORSORGER
Replacing [[-1, 0]] to NaN in Attribute GEBAEUDETYP
Replacing [[-1, 0]] to NaN in Attribute GEOSCORE_KLS7
Replacing [[-1

Replacing [[-1]] to NaN in Attribute KBA13_KW_80
Replacing [[-1]] to NaN in Attribute KBA13_KW_90
Replacing [[-1]] to NaN in Attribute KBA13_KW_110
Replacing [[-1]] to NaN in Attribute KBA13_KW_120
Replacing [[-1]] to NaN in Attribute KBA13_KW_121
Replacing [[-1]] to NaN in Attribute KBA13_MAZDA
Replacing [[-1]] to NaN in Attribute KBA13_MERCEDES
Replacing [[-1]] to NaN in Attribute KBA13_MOTOR
Replacing [[-1]] to NaN in Attribute KBA13_NISSAN
Replacing [[-1]] to NaN in Attribute KBA13_OPEL
Replacing [[-1]] to NaN in Attribute KBA13_PEUGEOT
Replacing [[-1]] to NaN in Attribute KBA13_RENAULT
Replacing [[-1]] to NaN in Attribute KBA13_SEG_GELAENDEWAGEN
Replacing [[-1]] to NaN in Attribute KBA13_SEG_GROSSRAUMVANS
Replacing [[-1]] to NaN in Attribute KBA13_SEG_KLEINST
Replacing [[-1]] to NaN in Attribute KBA13_SEG_KLEINWAGEN
Replacing [[-1]] to NaN in Attribute KBA13_SEG_KOMPAKTKLASSE
Replacing [[-1]] to NaN in Attribute KBA13_SEG_MINIVANS
Replacing [[-1]] to NaN in Attribute KBA13_SEG_MIN

### Missing values

Now that we have replaced all the unkowns to np.nan's we can now look at how many missing values each column contains and decide whether to keep a column for further analysis.

In [106]:
def get_missing_report(df):
    """
    Takes in a dataframe and returns a dataframe with information about column-wise missing values percentages.
    """
    missing_percen = df.isna().sum() * 100/ len(df)
    
    missing_percen_df = pd.DataFrame({"Attribute": df.columns,
                                     "Missing_Percentage": missing_percen}).reset_index(drop=True)
    return missing_percen_df

In [108]:
customer_missing_report = get_missing_report(customers)
azdias_missing_report = get_missing_report(azdias)

In [116]:
print("Total No of columns in Customers - ", len(customers.columns))
print("No of columns with missing values in customers - ",
      len(customer_missing_report[customer_missing_report.Missing_Percentage > 0]))

print("Total No of columns in Azdias - ", len(azdias.columns))
print("No of columns with missing values in azdias - ",
      len(azdias_missing_report[azdias_missing_report.Missing_Percentage > 0]))

Total No of columns in Customers -  369
No of columns with missing values in customers -  279
Total No of columns in Azdias -  366
No of columns with missing values in azdias -  279


## Part 1: Customer Segmentation Report

The main bulk of your analysis will come in this part of the project. Here, you should use unsupervised learning techniques to describe the relationship between the demographics of the company's existing customers and the general population of Germany. By the end of this part, you should be able to describe parts of the general population that are more likely to be part of the mail-order company's main customer base, and which parts of the general population are less so.

## Part 2: Supervised Learning Model

Now that you've found which parts of the population are more likely to be customers of the mail-order company, it's time to build a prediction model. Each of the rows in the "MAILOUT" data files represents an individual that was targeted for a mailout campaign. Ideally, we should be able to use the demographic information from each individual to decide whether or not it will be worth it to include that person in the campaign.

The "MAILOUT" data has been split into two approximately equal parts, each with almost 43 000 data rows. In this part, you can verify your model with the "TRAIN" partition, which includes a column, "RESPONSE", that states whether or not a person became a customer of the company following the campaign. In the next part, you'll need to create predictions on the "TEST" partition, where the "RESPONSE" column has been withheld.

In [ ]:
mailout_train = pd.read_csv('../../data/Term2/capstone/arvato_data/Udacity_MAILOUT_052018_TRAIN.csv', sep=';')

## Part 3: Kaggle Competition

Now that you've created a model to predict which individuals are most likely to respond to a mailout campaign, it's time to test that model in competition through Kaggle. If you click on the link [here](http://www.kaggle.com/t/21e6d45d4c574c7fa2d868f0e8c83140), you'll be taken to the competition page where, if you have a Kaggle account, you can enter. If you're one of the top performers, you may have the chance to be contacted by a hiring manager from Arvato or Bertelsmann for an interview!

Your entry to the competition should be a CSV file with two columns. The first column should be a copy of "LNR", which acts as an ID number for each individual in the "TEST" partition. The second column, "RESPONSE", should be some measure of how likely each individual became a customer – this might not be a straightforward probability. As you should have found in Part 2, there is a large output class imbalance, where most individuals did not respond to the mailout. Thus, predicting individual classes and using accuracy does not seem to be an appropriate performance evaluation method. Instead, the competition will be using AUC to evaluate performance. The exact values of the "RESPONSE" column do not matter as much: only that the higher values try to capture as many of the actual customers as possible, early in the ROC curve sweep.

In [ ]:
mailout_test = pd.read_csv('../../data/Term2/capstone/arvato_data/Udacity_MAILOUT_052018_TEST.csv', sep=';')